### ISSUE TITLE
Optimize executeBatch in JDBC driver

### ISSUE THREAD

**Issue by ebyhr** [prestosql/presto#5913]
#5507 implemented `PreapredStatement.executeBatch` in JDBC driver, but it will be sent each statements to the server. 

Related to #1777 & #5507

**Comment by findepi** [comment]
@martint do you have any thoughts how this could look like from Presto parser/analyzer perspective?

**Comment by dain** [comment]
I wonder if this can be rewritten to a JOIN with a VALUES expression.  Clearly it wouldn't work for every query, but where it didn't work, I wouldn't expect `executeBatch` to work anyway.

BTW, I think what people really want here it to be able to stream results into a table.  Something like postgres https://www.postgresql.org/docs/9.4/sql-copy.html, where you are basically creating a remote table in the client and having the Presto cluster integrate that into a query.  This is clearly much more work than executing a query for each batch of rows.

**Comment by findepi** [comment]
@dain i agree. We should have better higher level options than just "do a bunch of inserts".
Yet, i think we could improve inserts as a common primitive scenario that could just work better (somehow).

> I wonder if this can be rewritten to a JOIN with a VALUES expression.

We could easily rewrite batched INSERT with single rows as single INSERT with bigger VALUES, but this it not entirely correct:

- we can't just "rewrite query" because we need to be able to report errors referring to original query and it's textual locations
- big VALUES can have different semantics then series of singular INSERTs when bound parameters turn out to be different types -- the VALUES gets common super type which then get cast to table's type, whereas in series of INSERTs, each rows gets cast directly. This could lead to differences in results
  - example, inserting BIGINT 2^62+3 and DOUBLE 1.0 into decimal(38,0) column. Row-by-row the BIGINT value gets preserved, while with single big VALUES, it's gets converted into double with information loss.

**Comment by dain** [comment]
@findepi I understand.  I'm saying that unless it works out **exactly correctly** then you either fail or fall back to record at a time.

**Comment by dain** [comment]
@walter-weinmann I understand.  What I'm attempting to say, poorly, is that, I skeptical doing this as batch together a bunch of inserts will result in the performance you are seeking.  To do that, I think we need to treat the input data as a table in Presto, and we would stream these results from the client.  This is similar to the normal solution we suggest to people asking about this which is:
1. copy the file to HDFS/S3/some object store
2. add an external table in the hive connector
3. execute insert into by selecting from the table
4. drop the external table
5. delete the file

I don't expect that to work for you situation, I'm just presenting it as an example.

**Comment by ztaylor797** [comment]
Hi there, is this item on the roadmap? The interim solution above is, well, not very easy to implement for our application and, in general, loses a lot of the benefit of abstracting S3 access away from the application and onto Trino.

I admit it also took us some time to realize this was the problem with our slow jdbc-client hive inserts, partly due to the fact that TrinoDatabaseMetaData#supportsBatchUpdates returns true.

Support for this functionality would be very useful and much appreciated.

**Comment by davseitsev** [comment]
Our use case: we insert data from Airflow python jobs into Trino iceberg tables. According to our security requirements we can't write data directly to S3 or any other storage, so we generate INSERT queries and run them on Trino. 

I tested different approaches:
* ordinary `insert .. select .. from values` statement is slow, 12K records are inserted in 26 seconds (1 batch)
* prepared `insert .. select .. from values` statement is super slow: 1 prepare request for 1k rows and 12 executions 1000 rows in each take 60 seconds.
* `insert .. select .. from cast(parse_json(<json>) as array(row(.....)))` is the fastest way: 12K rows are inserted in 1 second.
There could be some difficulties with types but we managed to handle them. 
Maybe it would be useful for someone as a temporary workaround.

**Comment by o-shevchenko** [comment]
We are hitting this issue in practice. We are building a generic JDBC connector that moves data between systems, and executeBatch() performance is a major bottleneck.

We've confirmed that a JSON+UNNEST approach is 3-4x faster than standard JDBC batching, which proves the engine can handle the throughput the limitation is in the client protocol sending one statement per roundtrip.

However, we cannot use the JSON workaround in production. Since we are building a generic connector, we don't know types and schemas ahead of time. Maintaining correct serialization for all data types (temporal precision, decimal scale, varbinary/base64, identifier escaping, etc.) is brittle and unsustainable.

We would like to rely on the JDBC driver itself to provide efficient bulk insert support, rather than working around it at the application level.

More details and benchmarks in #28462.

@ebyhr Do you know if such a work is planned?
Thanks!
